anndata/readwrite/h5ad.py

```diff
@@ -0,0 +1,412 @@
+        if not (adata.isbacked and Path(adata.filename) == Path(filepath)):
+            # Otherwise, X should already be up to date
+            write_attribute(f, "X", adata.X, dataset_kwargs)
+        write_attribute(f, "obs", adata.obs, dataset_kwargs)
+        write_attribute(f, "var", adata.var, dataset_kwargs)
+        write_attribute(f, "obsm", adata.obsm, dataset_kwargs)
+        write_attribute(f, "varm", adata.varm, dataset_kwargs)
+        write_attribute(f, "layers", adata.layers, dataset_kwargs)
+        write_attribute(f, "uns", adata.uns, dataset_kwargs)
+        write_attribute(f, "raw", adata.raw, dataset_kwargs)
+    if adata.isbacked:
+        adata.file.open(filepath, 'r+')
+
+
+def _write_method(cls: Type[T]) -> Callable[[H5Group, str, T], None]:
+    return _find_impl(cls, H5AD_WRITE_REGISTRY)
+
+
+def write_attribute(f: H5Group, key: str, value, dataset_kwargs: Mapping):
+    if key in f:
+        del f[key]
+    _write_method(type(value))(f, key, value, dataset_kwargs)
+
+
+def write_raw(f, key, value, dataset_kwargs={}):
+    group = f.h5f.create_group(key)
+    group.attrs["encoding-type"] = "raw"
+    group.attrs["encoding-version"] = "0.1.0"
+    group.attrs["shape"] = value.shape
+    write_attribute(f, "raw/X", value.X, dataset_kwargs)
+    write_attribute(f, "raw/var", value.var, dataset_kwargs)
+    write_attribute(f, "raw/varm", value.varm, dataset_kwargs)
+
+
+def write_not_implemented(f, key, value, dataset_kwargs={}):
+    # If it's not an array, try and make it an array. If that fails, pickle it.
+    # Maybe rethink that, maybe this should just pickle, and have explicit implementations for everything else
+    raise NotImplementedError(
+        f"Failed to write value for {key}, since a writer for type {type(value)}"
+        f" has not been implemented yet."
+    )
+
+
+def write_basic(f, key, value, dataset_kwargs={}):
+    f.create_dataset(key, value, **dataset_kwargs)
+
+
+def write_list(f, key, value, dataset_kwargs={}):
+    write_array(f, key, np.array(value), dataset_kwargs)
+
+
+def write_none(f, key, value, dataset_kwargs={}):
+    pass
+
+
+def write_scalar(f, key, value, dataset_kwargs={}):
+    if "compression" in dataset_kwargs:  # Can't compress scalars, error is thrown
+        dataset_kwargs = dataset_kwargs.copy()
+        dataset_kwargs.pop("compression")
+    write_array(f, key, np.array(value), dataset_kwargs)
+
+
+def write_array(f, key, value, dataset_kwargs={}):
+    # Convert unicode to fixed length strings
+    if value.dtype.kind in {'U', 'O'}:
+        value = value.astype(adh5py.special_dtype(vlen=str))
+    elif value.dtype.names is not None:
+        value = _to_hdf5_vlen_strings(value)
+    f.create_dataset(key, data=value, **dataset_kwargs)
+
+
+def write_dataframe(f, key, df, dataset_kwargs={}):
+    if isinstance(f, adh5py.File):  # If it's the patched one
+        group = f.h5f.create_group(key)
+    else:
+        group = f.create_group(key)
```

Comment:
Added. In general, I would like to remove the `anndata.h5py` file, since I'm pretty sure we can handle sparse matrices without also having our own `File`, `Group`, and `Dataset`. Since we already have to deal with indexing ourselves, I think it'll be easier to share code between backends (like `zarr`) as well. Do have any thoughts on this?

Thanks! Now I just need to actually use it.